File: computer3/Winter Term/8 - Mar 05/lvl3_hw_sol/square_pool.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException {
		int n = readInt(), t = readInt(), x[]= new int[t+2], y[] = new int[t+2];
		for(int i=0; i<t; i++) {
			x[i] = readInt(); y[i] = readInt();
		}
		t+=2;  x[t-1]=n+1; y[t-1]=n+1;
		int ans = 0;
		for(int i=0; i<t; i++) {
			for(int j=i+1; j<t; j++) {
				int lft = Math.min(x[i], x[j]), rit=Math.max(x[i],x[j]);
				int width = rit - lft - 1;
				List<Integer> mid = new ArrayList<>();
				for(int k=0; k<t; k++) {
					if(x[k] > lft && x[k] < rit) mid.add(y[k]);
				}
				mid.add(0); mid.add(n+1);
				Collections.sort(mid);
				for(int k=1; k<mid.size(); k++) {
					int height = mid.get(k) - mid.get(k-1) - 1;
					ans = Math.max(ans,Math.min(width, height));
				}
			}
		}
		System.out.println(ans);
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: computer3/Winter Term/8 - Mar 05/lvl3_hw_sol/square_pool.cpp
#include <bits/stdc++.h>
using namespace std;
typedef pair<int, int> pi;
int N, T, ans; vector<pi> a;
int main(){
    cin >> N >> T;
    for(int i=1, x, y; i<=T; i++) {
        cin >> x >> y; a.push_back({x, y});
    }
    a.push_back({0, 0}); a.push_back({N+1, N+1});
    for(int i=0; i<a.size(); i++){
        for(int j=i+1; j<a.size(); j++){
            int lft = min(a[i].first, a[j].first), rit = max(a[i].first, a[j].first);
            int wid = rit - lft - 1;
            vector<int> mid = {0, N+1};
            for(int k=0; k<a.size(); k++){
                if(a[k].first > lft && a[k].first < rit){
                    mid.push_back(a[k].second);
                }
            }
            sort(mid.begin(),  mid.end());
            for(int k=1; k<mid.size(); k++){
                int height = mid[k] - mid[k-1] - 1;
                ans = max(ans, min(wid, height));
            }
        }
    }
    cout << ans << endl;
}